ports/rp2/machine_uart.c

```diff
@@ -179,31 +212,52 @@ STATIC const mp_rom_map_elem_t machine_uart_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_write), MP_ROM_PTR(&mp_stream_write_obj) },
 
     { MP_ROM_QSTR(MP_QSTR_sendbreak), MP_ROM_PTR(&machine_uart_sendbreak_obj) },
+
+    { MP_ROM_QSTR(MP_QSTR_INV_TX), MP_ROM_INT(UART_INVERT_TX) },
+    { MP_ROM_QSTR(MP_QSTR_INV_RX), MP_ROM_INT(UART_INVERT_RX) },
+
 };
 STATIC MP_DEFINE_CONST_DICT(machine_uart_locals_dict, machine_uart_locals_dict_table);
 
 STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
     machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    // TODO support timeout
+    uint64_t t = time_us_64() + (uint64_t)self->timeout * 1000;
```

Comment:
I don't think it's necessary to use uint64_t for the timeout, a uint32_t should be plenty, and use `mp_hal_ticks_ms()` to get the current timestamp?

Reply:
I used uint64_t for the us timer, to avoid overflow errors. The time_us_64() call is very lightweight. It just accesses two registers. mp_hal_ticks_ms() goes through several call levels and calls in the end    time_us_64() too, including a division by 1000 instead of a multiplication. So calling time_us_64() directly costs less cycles. Using mp_hal_ticks_ms() would have a style advantage, being in the mp_xxxxx class of support functions.

Reply:
Ok, good reasoning.  I was just concerned about 64-bit arithmetic but it's probably quicker than `mp_hal_ticks_ms()`.  And a 64-bit us timer takes about 584,000 years to overflow so that's not an issue.